File: recap/ThreeSquare/ThreeSquare.java
import java.util.*;
public class ThreeSquare {
	public static void main(String[] args) {
		Scanner scan = new Scanner(System.in);
		int[][] grid = new int[3][2];
		int area = 0;
		for(int i = 0; i < 3; i++) {
			for(int j = 0; j < 2; j++) {
				grid[i][j] = scan.nextInt();
			}
			area += grid[i][0] * grid[i][1];
		}
		int side = (int)Math.sqrt(area);
		if(side * side != area) {
			System.out.println("NO");
		}
		else {
			System.out.println(can(new boolean[3], grid, side, side) ? "YES" : "NO");
		}
	}
	
	public static boolean can(boolean[] used, int[][] sides, int s1, int s2) {
		int amtAvail = 0;
		for(boolean out: used) {
			if(!out) {
				amtAvail++;
			}
		}
		if(amtAvail == 1) {
			for(int i = 0; i < sides.length; i++) {
				if(!used[i]) {
					return (sides[i][0] == s1 && sides[i][1] == s2) || (sides[i][0] == s2 && sides[i][1] == s1);
				}
			}
			throw new RuntimeException();
		}
		for(int i = 0; i < sides.length; i++) {
			if(!used[i]) {
				used[i] = true;
				if(sides[i][0] == s1 && sides[i][1] < s2 && can(used, sides, s1, s2 - sides[i][1])) {
					return true;
				}
				if(sides[i][0] == s2 && sides[i][1] < s1 && can(used, sides, s1 - sides[i][1], s2)) {
					return true;
				}
				if(sides[i][1] == s1 && sides[i][0] < s2 && can(used, sides, s1, s2 - sides[i][0])) {
					return true;
				}
				if(sides[i][1] == s2 && sides[i][0] < s1 && can(used, sides, s1 - sides[i][0], s2)) {
					return true;
				}
				used[i] = false;
			}
		}
		return false;
	}
	
}

File: recap/ThreeSquare/ThreeSquare.cpp
#include <iostream>
#include <vector>
#include <math.h>
using namespace std ;
int main(int argc, char *argv[]) {
   vector<int> v(6) ;
   for (int i=0; i<6; i++)
      cin >> v[i] ;
   int a = v[0]*v[1] + v[2]*v[3] + v[4]*v[5] ;
   int s = (int)sqrt(a) ;
   int success = 0 ;
   if (s * s == a) {
      int s2 = -1 ;
      for (int i=0; i<6; i += 2)
         if (v[i] == s || v[i+1] == s) {
            swap(v[4], v[i]) ;
            swap(v[5], v[i+1]) ;
            s2 = 2 * s - v[4] - v[5] ;
         }
      if (s2 > 0) {
         if (((v[0] == s || v[1] == s) && (v[2] == s || v[3] == s)) ||
             ((v[0] == s2 || v[1] == s2) && (v[2] == s2 || v[3] == s2)))
            success++ ;
      }
   }
   cout << (success ? "YES" : "NO") << endl ;
}
